### DISCUSSION TITLE
WPF有没有控件可以取代winform的picturebox？

### DISCUSSION THREAD

**Discussion by N4v1zzZ** [BYJRK/DotNet-Discussions#68] [WPF 开发]
一个历史遗留问题，工作上有项目要使用海康威视的摄像头，几年前写的winform程序里，调用摄像头的sdk仅需页面上放一个picturebox然后传入句柄（handler）即可显示视频流。
现在用WPF重构，没有太好的替代控件，但是又不太想再WPF下继续使用winform控件，不知道各位朋友有没有更好的办法？

**Comment by BeautyOfCoding** [comment]
我也一直被这个问题困惑着，貌似国内一些常用的相机厂商，比如海康，大华等等提供的SDK在预览的时候都需要提供控件句柄，但是WPF只有窗口句柄，没有控件句柄的概念，只能引入Winform的picturebox，感觉好麻烦啊

**Reply by spikeOO7** [reply]
我用的是Image 控件 实现视频流，为了使用MVVM，中间加了图像转换。

**Reply by spikeOO7** [reply]
<Image HorizontalAlignment="Left" Height="300" Margin="10,10,0,0" VerticalAlignment="Top" Width="509" x:Name="imageDis" Source="{Binding ImageDisplay}"/>


private BitmapImage imageDisplay;
public BitmapImage ImageDisplay { get => imageDisplay; set => SetProperty(ref imageDisplay, value); }

private void ImageCallbackFunc(IntPtr pData, ref MyCamera.MV_FRAME_OUT_INFO_EX pFrameInfo, IntPtr pUser)
 {
        DateTime dt = DateTime.Now;

        MyCamera.MV_PIXEL_CONVERT_PARAM stConverPixelParam = new MyCamera.MV_PIXEL_CONVERT_PARAM();
        Bitmap bitmap = BackGray(pData, pFrameInfo, ref stConverPixelParam);

        BitmapImage image = BitmapToBitmapImage(bitmap);
        TimeOutGet = (DateTime.Now - dt).TotalMilliseconds;//获取一帧图像时间（）

        ImageDisplay = image;

        TimeOutDisplay = (DateTime.Now - dt).TotalMilliseconds;//显示时间
}


private Bitmap BackGray(IntPtr pData, MyCamera.MV_FRAME_OUT_INFO_EX pFrameInfo, ref MyCamera.MV_PIXEL_CONVERT_PARAM stConverPixelParam)
{
    try
    {
        Int32 nWidth = pFrameInfo.nWidth;
        Int32 nHeight = pFrameInfo.nHeight;
        Int32 lenData = nWidth * nHeight * 1; //数据长度，8位图像数据的长度=宽度*高度*1
        byte[] buffer = new byte[lenData]; //创建指定长度byte数据
        IntPtr pTemp = IntPtr.Zero;
        Bitmap bmp = null;
        MyCamera.MvGvspPixelType enDstPixelType = MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono8;
        lock (this)
        {
            if (IsMonoData(pFrameInfo.enPixelType))
            {
                enDstPixelType = MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono8;
                nSaveImageNeedSize = (uint)pFrameInfo.nWidth * pFrameInfo.nHeight;
            }
            else if (IsColorData(pFrameInfo.enPixelType))
            {
                enDstPixelType = MyCamera.MvGvspPixelType.PixelType_Gvsp_BGR8_Packed;
                nSaveImageNeedSize = (uint)pFrameInfo.nWidth * pFrameInfo.nHeight * 3;
            }

            pTemp = pData;

            if (m_nBufSizeForSaveImage < nSaveImageNeedSize)
            {
                if (m_BufForSaveImage != IntPtr.Zero)
                {
                    Marshal.Release(m_BufForSaveImage);
                }
                m_nBufSizeForSaveImage = nSaveImageNeedSize;
                m_BufForSaveImage = Marshal.AllocHGlobal((Int32)m_nBufSizeForSaveImage);
            }
            stConverPixelParam.nWidth = pFrameInfo.nWidth;
            stConverPixelParam.nHeight = pFrameInfo.nHeight;
            stConverPixelParam.pSrcData = pData;
            stConverPixelParam.nSrcDataLen = pFrameInfo.nFrameLen;
            stConverPixelParam.enSrcPixelType = pFrameInfo.enPixelType;
            stConverPixelParam.enDstPixelType = enDstPixelType;
            stConverPixelParam.pDstBuffer = m_BufForSaveImage;
            stConverPixelParam.nDstBufferSize = m_nBufSizeForSaveImage;
            //m_MyCamera.MV_CC_ConvertPixelType_NET(ref stConverPixelParam);
            pTemp = m_BufForSaveImage;
        }

        lock (BufForDriverLock)
        {
            if (enDstPixelType == MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono8)
            {
                //************************Mono8 转 Bitmap*******************************

                //test
                //Bitmap bmp99 = new Bitmap("Image_20221210161714083.bmp");
                //BitmapData bmpData = bmp99.LockBits(new Rectangle(0, 0, bmp99.Width, bmp99.Height), ImageLockMode.ReadOnly, bmp99.PixelFormat);
                //Bitmap bmp99New = new Bitmap(200, 200, bmpData.Stride, bmp99.PixelFormat, bmpData.Scan0);
                //bmp99.UnlockBits(bmpData);
                //bmp99.Save("99new.bmp");

                bmp = new Bitmap(pFrameInfo.nWidth, pFrameInfo.nHeight, pFrameInfo.nWidth, PixelFormat.Format8bppIndexed, pData);
                ColorPalette cp = bmp.Palette;
                // init palette
                for (int i = 0; i < 256; i++)
                {
                    cp.Entries[i] = Color.FromArgb(i, i, i);
                }
                // set palette back
                bmp.Palette = cp;

                //bmp.Save("image.bmp", ImageFormat.Bmp);
            }

            else
            {
                //*********************BGR8 转 Bitmap**************************
                try
                {
                    unsafe
                    {
                        byte* bitmap = (byte*)pTemp.ToPointer();
                        bmp = new Bitmap(pFrameInfo.nWidth, pFrameInfo.nHeight, pFrameInfo.nWidth * 1, PixelFormat.Format8bppIndexed, pTemp);
                        BitmapData bitmapData = bmp.LockBits(new Rectangle(0, 0, pFrameInfo.nWidth, pFrameInfo.nHeight), ImageLockMode.ReadWrite, bmp.PixelFormat);
                    }


                    //bmp.Save("image.bmp", ImageFormat.Bmp);
                }
                catch
                {
                    //ShowErrorMsg("Write File Fail!", 0);
                }
            }

        }
        return bmp;

    }
    catch (Exception)
    {

        throw;
    }

}


private BitmapImage BitmapToBitmapImage(System.Drawing.Bitmap bitmap)
{

    using (System.IO.MemoryStream ms = new System.IO.MemoryStream())
    {
        bitmap.Save(ms, ImageFormat.Bmp);

        ms.Position = 0;
        BitmapImage bitmapImage = new BitmapImage();
        bitmapImage.BeginInit();
        bitmapImage.CacheOption = BitmapCacheOption.OnLoad;
        bitmapImage.StreamSource = ms;
        bitmapImage.EndInit();
        bitmapImage.Freeze();
        return bitmapImage;
    }
}

**Comment by N4v1zzZ** [comment]
没人能回答嘛

**Comment by SalHe** [comment]
有没有考虑外挂一个窗口？

**Comment by LiuChao19940201** [comment]
直接用image控件就行，我就是这么用的


![image](https://github.com/BYJRK/DotNet-Discussions/assets/145536444/ba59a630-701e-4361-9bdf-fd80e2367c3a)
![image](https://github.com/BYJRK/DotNet-Discussions/assets/145536444/11305544-b96b-431f-a481-f1ffa6441427)

**Reply by N4v1zzZ** [reply]
ViewModel里是怎么实现的？

**Reply by LiuChao19940201** [reply]
![image](https://github.com/BYJRK/DotNet-Discussions/assets/145536444/e9f48b03-0374-4b27-b7a5-110a1ae6b3db)
通过相机的SDK获取到图形的对象，然后直接给image的数据源赋值就可以了
